Topics/Tech_Stacks/cuda.md

```diff
@@ -0,0 +1,260 @@
+
+Another important topic that will come up is GPU memory. Often referred to as device memory (whereas RAM is considered host memory), GPUs have built-in memory to avoid the slow bandwidth encountered with traditional RAM that the CPU communicates with. This GPU memory also referred to as VRAM is high-speed, high-throughput volatile memory that is meant for high-bandwidth operations. When executing GPU code, note that we must transfer elements such as arrays from traditional RAM onto our GPU memory.
+
+The following images below will help clear up any confusion about our GPU architecture.
+
+![image](https://docs.nvidia.com/cuda/cuda-c-programming-guide/_images/gpu-devotes-more-transistors-to-data-processing.png)
+
+![image](https://docs.nvidia.com/cuda/cuda-c-programming-guide/_images/automatic-scalability.png)
+
+Source: https://docs.nvidia.com/cuda/cuda-c-programming-guide/
+
+## Writing our first kernel:
+
+Note that the CUDA language is effectively a wrapper for C++, with additional runtime libraries provided by NVIDIA. This guide assumes that you are somewhat comfortable with C or C++. Lets start by launching a kernel, which is a specialized function that is executed in parallel by multiple threads on the GPU. In CUDA, this involves defining the kernel with the `__global__` qualifier, indicating it's meant to be run on the GPU. When kernels are launched, they spawn a large number of threads specified in the angular brackets like so, with the block size, and the threads per block being specified like so: `<<<blockSize, threadSize>>>example_kernel(param1, param2, ...)`. Lets define a kernel that multiplies two arrays together and puts the result in another array. But first, we'll need to allocate some device memory. See the following:
+
+```cpp
+#include <iostream>
+__global__ void multiply_floats(float *a, float *b, float *result, int n) {
+    int index = threadIdx.x + blockIdx.x * blockDim.x;
+    // Get the global thread index by calculating which block and thread inside each block we're in
+
+    // Make sure we don't calculate over the array size
+    if (index < n) {
+        result[index] = a[index] * b[index];
+    }
+}
+
+int main() {
+    int n = 100000000; // Size of the arrays
+    float *a, *b, *result; // Pointers for host (CPU)arrays
+    float *d_a, *d_b, *d_result; // Pointers for device (GPU) arrays
+    size_t size = n * sizeof(float);
+
+    // Allocate memory on the host
+    a = (float*)malloc(size);
+    b = (float*)malloc(size);
+    result = (float*)malloc(size);
+
+    // Allocate memory on the device
+    cudaMalloc(&d_a, size);
+    cudaMalloc(&d_b, size);
+    cudaMalloc(&d_result, size);
+
+    // Copy data from host to device
+    cudaMemcpy(d_a, a, size, cudaMemcpyHostToDevice);
```

Comment:
`cudaMemcpyHostToDevice` appears to be undefined

Reply:
I should probably mention that `cudaMemcpyHostToDevice` is a constant defined in the CUDA runtime library.